File: ConstructBSTfromInorderPostorder.java

/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */
class ConstructBSTFromInorderPostorder {

    HashMap<Integer, Integer> map;
    int idx;

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if (postorder.length == 0)
            return null;
        map = new HashMap<>();
        idx = postorder.length - 1;
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }
        return helper(postorder, 0, inorder.length - 1);

    }

    private TreeNode helper(int[] postorder, int start, int end) {
        if (start > end) {
            return null;
        }
        TreeNode root = new TreeNode(postorder[idx]);
        int rootIdx = map.get(postorder[idx]);
        idx--;
        root.right = helper(postorder, rootIdx + 1, end);
        root.left = helper(postorder, start, rootIdx - 1);
        return root;
    }
}

// TC: O(n) SC: O(n)

File: ConstructBSTfromInorderPostorder.py
# Brute Force DFS solution

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        
        if len(inorder) == 0 or len(postorder) == 0 or postorder == None:
            return
        
        root_val = postorder.pop()
        root = TreeNode(root_val)
        
        root_index = inorder.index(root.val) #O(n) time
        
    
        
        root.right = self.buildTree(inorder[root_index + 1:], postorder) # O(n) time and space
        root.left = self.buildTree(inorder[:root_index], postorder) # O(n) time and space
        
        return root
    
    # Time Complexity: O(n^2)
    # Space Complexity: O(n^2)

# Time and space efficient solution using Hashmap

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        
        hashmap = {}
        for i in range(len(inorder)):
            hashmap[inorder[i]] = i
        def helper(left, right):
            
            if left > right:
                return
            root = TreeNode(postorder.pop())
            index = hashmap[root.val]
            root.right = helper(index + 1, right)
            root.left = helper(left, index-1)
            return root
        
        return helper(0, len(inorder)-1)
    
     # Time Complexity: O(n)
     # Space Complexity: O(n)